File: src/main/python/algo/sort/merge_sort.py
# Uncomment print statements to understand the concept more clearly
def merge_sort(list):
    """
    Sorts a list in ascending order
    - Divide: Find the midpoint of the list and divide into subsets
    - Conquer: Recursively sort the sublists created in previous step
    - Combine: Merge the sorted sublists created in previous step

    uses split(list) and merge(left_list, right_list) as helper methods

    Returns a new sorted list
    Takes Overall O(kn log n) time
    """

    # Naively sorted when the given list is empty or with length of 1
    if len(list) <= 1:
        return list

    # two sublists are returned and captured in two variables
    left_half, right_half = split(list)

    # print(f"left half = {left_half}")
    left = merge_sort(left_half)
    # print(f"left = {left}")
    # print(f"right half = {right_half}")
    right = merge_sort(right_half)
    # print(f"right = {right}")

    return merge(left, right)


def split(list):
    """
    Divide the unsorted list at midpoint into sublists

    Returns two sublists - left and right

    slicing operation takes O(k) time, where k is the slized size
    Takes Overall O(k log n) time
    """

    mid = len(list) // 2  # floor division (//) to find the mid point
    left = list[:mid]  # slicing from start to mid(exclusive)
    right = list[mid:]  # slicing from mid to end

    return left, right


def merge(left, right):
    """
    Merges two lists, sorting them in process

    Returns a new merged list
    Takes O(n) time or Linear Time
    """

    sorted_values = []
    left_index = 0
    right_index = 0

    # compare the values from left list with the values from right list
    # exit the loop when either of the indexes reaches their respective list lengths
    while left_index < len(left) and right_index < len(right):
        # print(f"left = {left[left_index]} - right = {right[right_index]}")
        if left[left_index] < right[right_index]:
            sorted_values.append(left[left_index])
            left_index += 1
            # print(f"left index = {left_index}")
        else:
            sorted_values.append(right[right_index])
            right_index += 1
            # print(f"right index = {right_index}")

    # when the left list is larger than the right list. assuming the values inside the list is already sorted
    while left_index < len(left):
        # print(f"left = {left[left_index]} < len")
        sorted_values.append(left[left_index])
        left_index += 1
        # print(f"left index = {left_index}")

    # when the right list is larger than the left list. assuming the values inside the list is already sorted
    while right_index < len(right):
        # print(f"right = {right[right_index]} < len")
        sorted_values.append(right[right_index])
        right_index += 1
        # print(f"right index = {right_index}")

    # print(f"sorted values = {sorted_values}")
    return sorted_values


def verify_sorted(list):
    n = len(list)

    if n == 0 or n == 1:
        return True

    # recursively check the list
    # compare the first two elements in the list
    # recursively pass sublist from the index 1 to the end of the sublist
    return list[0] < list[1] and verify_sorted(list[1:])


a_list = [50, 43, 8, 123, 9, 4, 87, 3]
l = merge_sort(a_list)

print(a_list)
print(verify_sorted(a_list))

print(l)
print(verify_sorted(l))

File: src/main/algo/sort/merge/MergeSort.java
package main.algo.sort.merge;

public class MergeSort {

    // Time Complexity: O(n log n)
    // Space Complexity: O(n)
    public static void mergeSort(int[] arr) {
        mergeSortHelper(arr, 0, arr.length - 1);
    }

    private static void mergeSortHelper(int[] arr, int low, int high) {
        // sublist at least has 2 elements
        if (low < high) {
            int midIndex = low + ((high - low) / 2); // (low + high) / 2
            mergeSortHelper(arr, low, midIndex); // divide into sub lists (left side)
            mergeSortHelper(arr, midIndex + 1, high); // divide into sub lists (right side)
            merge(arr, low, midIndex, high); // conquer and combine the sub lists
        }
    }

    // Theta(m + n)
    private static void merge(int[] arr, int low, int mid, int high) {
        // get the size of the sublist
        int leftSize = mid - low + 1;
        int rightSize = high - mid;

        // create sublist and copy the contents to it
        int[] leftSublist = new int[leftSize];
        int[] rightSublist = new int[rightSize];

        // can use System.arraycopy() too
        for (int i = 0; i < leftSize; i++) {
            leftSublist[i] = arr[low + i];
        }
        for (int i = 0; i < rightSize; i++) {
            rightSublist[i] = arr[mid + 1 + i];
        }

        int leftIndex = 0;
        int rightIndex = 0;
        int placeIndex = low; // keep track of where to place the sorted element

        // compare and sort the left sublist with right sublist
        // exit when one of the sublist index reaches its end
        while (leftIndex < leftSize && rightIndex < rightSize) {
            if (leftSublist[leftIndex] <= rightSublist[rightIndex]) {
                arr[placeIndex] = leftSublist[leftIndex];
                leftIndex++;
            } else {
                arr[placeIndex] = rightSublist[rightIndex];
                rightIndex++;
            }
            placeIndex++;
        }

        // copy over the remaining content from a larger sublist
        while (leftIndex < leftSize) {
            arr[placeIndex] = leftSublist[leftIndex];
            leftIndex++;
            placeIndex++;
        }
        while (rightIndex < rightSize) {
            arr[placeIndex] = rightSublist[rightIndex];
            rightIndex++;
            placeIndex++;
        }

    }

    public static void printArray(int[] arr) {
        for (int item : arr) {
            System.out.print(item + " ");
        }
        System.out.println();
    }

    public static boolean isSorted(int[] arr) {
        int n = arr.length;

        for (int i = 0; i < n; i++) {
            if (arr[i + 1] == n) {
                break;
            }
            if (arr[i] > arr[i + 1]) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {

        int[] arr = {50, 43, 8, 123, 9, 4, 87, 3};
        printArray(arr);
        System.out.println(isSorted(arr));

        mergeSort(arr);
        printArray(arr);
        System.out.println(isSorted(arr));

    }

}
